backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/process/ModCntr_Specific_Delete_Price_Selection.java

```diff
@@ -80,29 +81,34 @@ public ProcessPreconditionsResolution checkPreconditionsApplicable(final @NonNul
 	}
 
 	@Override
+	@RunOutOfTrx
 	protected String doIt()
 	{
 		retrieveContractSpecificPricesFromSelection()
-				.forEach(contractPriceId ->
-				{
-					if (!modularContractPriceService.existsSimilarContractSpecificScalePrice(contractPriceId))
-					{
-						throw new AdempiereException(ERROR_MSG_NO_FALLBACK_PRICE);
-					}
-
-					final ModCntrSpecificPrice contractPrice = modularContractPriceService.getById(contractPriceId);
-
-					// delete price
-					modularContractPriceService.deleteById(contractPriceId);
-
-					// the update price by recomputing the price for logs
-					// the given price is ingonred in UserElementNumberShipmentLineLog
-					contractLogService.updatePriceAndAmount(ModCntrLogPriceUpdateRequest.builder()
-									.unitPrice(contractPrice.getProductPrice())
-									.flatrateTermId(contractPrice.flatrateTermId())
-									.modularContractModuleId(contractPrice.modularContractModuleId())
-									.build(),
-							logHandlerRegistry);
+				.forEach(contractPriceId -> {
+					trxManager.assertThreadInheritedTrxNotExists();
+					trxManager.runInThreadInheritedTrx(() ->
+							{
+								if (!modularContractPriceService.existsSimilarContractSpecificScalePrice(contractPriceId))
+								{
+									throw new AdempiereException(ERROR_MSG_NO_FALLBACK_PRICE);
```

Comment:
because u are handling each contract price in its own trx, 
if u are throwing an exception here, it means that all the previous deleted prices remain deleted.
We stop here on this one.

Not sure if that's the desired functionality.
Maybe you just want to send a notification for the problematic contract prices and go forward... not sure.

Reply:
Usually all contracts should be updated the same way, so if it fails it should already fail on the first contract. And even if not, the already deleted contracts shouldn't be part of the next run anymore, because the price matching the criteria is already deleted.